graph/pipeline.go

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package graph
+
+import "fmt"
+
+const (
+	// The default step timeout is 10 minutes.
+	defaultStepTimeoutInSeconds = 60 * 10
+
+	// The minimum step timeout is 10 seconds.
+	minStepTimeoutInSeconds = 10
+
+	// The maximum step timeout is 2 hours.
+	maxStepTimeoutInSeconds = 60 * 60 * 2
+
+	// The default total timeout is 1 hour.
+	defaultTotalTimeoutInSeconds = 60 * 60 * 1
+
+	// The minimum total timeout is 10 seconds.
+	minTotalTimeoutInSeconds = 10
+
+	// The maximum total timeout is 6 hours.
+	maxTotalTimeoutInSeconds = 60 * 60 * 6
+)
+
+// Pipeline represents a build pipeline.
+type Pipeline struct {
+	Steps        []*Step   `toml:"step"`
+	StepTimeout  int       `toml:"stepTimeout,omitempty"`
+	TotalTimeout int       `toml:"totalTimeout,omitempty"`
+	Push         []string  `toml:"push,omitempty"`
+	Secrets      []*Secret `toml:"secrets,omitempty"`
+	WorkDir      string    `toml:"workDir,omitempty"`
+}
+
+// NewPipeline returns a default Pipeline object.
+func NewPipeline(steps []*Step, push []string, secrets []*Secret) *Pipeline {
```

Comment:
>push []string, [](start = 32, length = 14)

is push still first-class? or replaced by docker push step?

Reply:
push isn't first class yet, more notably, it's not possible to multi-push w/ normalized image names from a pipeline yet. But it's a TODO